x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,206 @@
+		err = sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// validate the sender bech32 address from the counterparty chain
+	bech32Prefix := strings.Split(data.Sender, "1")[0]
+	if bech32Prefix == data.Sender {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender: %s", data.Sender).Error(),
+		)
+	}
+
+	senderBz, err := sdk.GetFromBech32(data.Sender, bech32Prefix)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender %s, %s", data.Sender, err.Error()).Error(),
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	if !sender.Equals(recipient) {
+		// return the original success acknowledgement since the address are different
+
+		// NOTE: here we can't know if the recipient is a 'secp256k1' address. If it is,
+		// then the user will need to withdraw the tokens using this same logic.
+		return ack
+	}
+
+	balances := k.bankKeeper.GetAllBalances(ctx, recipient)
+
+	// NOTE: since the balance is 0 we can reject the transaction
+	// We return error since the sender and recipient addresses are the same
+	if balances.IsZero() {
+		logger.Debug(
+			"rejected IBC transfer to 'secp256k1' key address",
+			"sender", data.Sender, "recipient", data.Receiver,
+		)
+
+		// return the an error acknowledgement since the address the same
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(types.ErrKeyTypeNotSupported, "receiver address %s is not a valid ethereum address", data.Receiver).Error(),
+		)
+	}
+
+	// swap the sender and recipient
+	sender = recipient
+
+	// transfer the balance back to the sender address
+	var (
+		recipientStr        string
+		srcPort, srcChannel string
+	)
+
+	for _, coin := range balances {
+		// we only transfer IBC tokens back to their respective source chains
+		if strings.HasPrefix(coin.Denom, "ibc/") {
+			// TODO: recipient should have the bech32 prefix from the chain
+			recipientStr = data.Sender
+
+			srcPort, srcChannel, err = k.GetIBCDenomSource(ctx, coin.Denom, data.Sender)
+			if err != nil {
+				return channeltypes.NewErrorAcknowledgement(err.Error())
+			}
+
+		} else {
+			// send Evmos native tokens to the source port and channel
+			recipientStr = data.Sender
+
+			srcPort = packet.SourcePort
+			srcChannel = packet.SourceChannel
+		}
+
+		// TODO: get the correct bech32 address of the source chain
+		recipientStr = data.Sender
```

Comment:
Yes, how would you be able to do this for arbitrary chains?

My suggestion is that you just do it on a per-chain basis.

If you want to retrieve your osmosis balance, send some dust from osmosis and we'll send your osmosis balance back. Same concept for other chains like Hub, Juno, etc.